nkms/blockchain/eth/sol/source/contracts/MinersEscrow.sol

```diff
@@ -305,48 +228,64 @@ contract MinersEscrow is Issuer {
         require(_value != 0 || _periods != 0);
+        mint();
 
-        uint256 lockedTokens = calculateLockedTokens(msg.sender, 1);
+        uint256 lockedTokens = getLockedTokens(msg.sender, 1);
         MinerInfo storage info = minerInfo[msg.sender];
         require(_value <= token.balanceOf(address(this)) &&
-            _value <= info.value.sub(lockedTokens));
+            _value <= info.value.sub(lockedTokens) &&
+            _value >= minAllowableLockedTokens &&
+            _value.add(lockedTokens) <= maxAllowableLockedTokens &&
+            _periods >= minLockedPeriods);
 
-        if (lockedTokens == 0) {
-            require(_value >= minAllowableLockedTokens);
-            info.lockedValue = _value;
-            info.maxReleasePeriods = Math.max256(_periods, minReleasePeriods);
-            info.releaseRate = Math.max256(_value.divCeil(info.maxReleasePeriods), 1);
-            info.release = false;
-        } else {
-            info.lockedValue = lockedTokens.add(_value);
-            info.maxReleasePeriods = info.maxReleasePeriods.add(_periods);
-            info.releaseRate = Math.max256(
-                info.lockedValue.divCeil(info.maxReleasePeriods), info.releaseRate);
-        }
-        require(info.lockedValue <= maxAllowableLockedTokens);
+        uint256 currentPeriod = getCurrentPeriod();
+        info.stakes.push(StakeInfo(currentPeriod.add(uint256(1)), currentPeriod.add(_periods), _value));
 
-        confirmActivity(info.lockedValue);
-        Locked(msg.sender, info.lockedValue, info.releaseRate);
-        mint();
+        confirmActivity(_value + lockedTokens, _value);
+        emit Locked(msg.sender, _value, currentPeriod + 1, currentPeriod + _periods);
     }
 
     /**
-    * @notice Switch lock
+    * @notice Divide stake into two parts
+    * @param _oldValue Old stake value
+    * @param _lastPeriod Last period of stake
+    * @param _newValue New stake value
+    * @param _periods Amount of periods for extending stake
     **/
-    function switchLock() public onlyTokenOwner {
+    function divideStake(
+        uint256 _oldValue,
+        uint256 _lastPeriod,
+        uint256 _newValue,
+        uint256 _periods
+    )
+        public onlyTokenOwner
+    {
+        require(_newValue >= minAllowableLockedTokens && _periods > 0);
         MinerInfo storage info = minerInfo[msg.sender];
-        info.release = !info.release;
-        LockSwitched(msg.sender, info.release);
+        for (uint256 index = 0; index < info.stakes.length; index++) {
+            StakeInfo storage stake = info.stakes[index];
+            if (stake.lockedValue == _oldValue &&
+                stake.lastPeriod == _lastPeriod) {
+                break;
+            }
+        }
+        // TODO lastPeriod can be equal current period but need to call confirmActivity
+        require(index < info.stakes.length && stake.lastPeriod >= getCurrentPeriod().add(uint256(1)));
```

Comment:
What is `index` equal to if we go through all the cycle without ever satisfying the `if` condition? Can we accidentally get a situation where wrong inputs is given for `_oldValue`, but still the stake which is selected has an index `info.stakes.length - 1`, instead of erroring out

Reply:
Ok, `require(index < info.stakes.length)` will raise if the cycle has ended, great

Reply:
About small stakes - old and new values are checked with minAllowableLockedTokens